torch/_inductor/codegen/segmented_tree.py

```diff
@@ -0,0 +1,248 @@
+        self.tree[node] = self.summary_op(self.tree[left_child], self.tree[right_child])
+
+    def _push_lazy(self, node: int, start: int, end: int) -> None:
+        """
+        Push lazy updates down to children.
+
+        Args:
+            node: Current node index
+            start: Start index of the segment
+            end: End index of the segment
+        """
+        lazy_node = self.lazy[node]
+        if lazy_node is not None:
+            # Apply lazy update to current node
+            self.tree[node] = self.update_op(self.tree[node], lazy_node)
+
+            if start != end:  # Not a leaf node
+                left_child = 2 * node
+                right_child = 2 * node + 1
+
+                # Propagate to children
+                lazy_left_child = self.lazy[left_child]
+                if lazy_left_child is None:
+                    self.lazy[left_child] = lazy_node
+                else:
+                    self.lazy[left_child] = self.update_op(lazy_left_child, lazy_node)
+
+                lazy_right_child = self.lazy[right_child]
+                if lazy_right_child is None:
+                    self.lazy[right_child] = lazy_node
+                else:
+                    self.lazy[right_child] = self.update_op(lazy_right_child, lazy_node)
+
+            # Clear the lazy value
+            self.lazy[node] = None
+
+    def _update_range_helper(
+        self, node: int, start: int, end: int, left: int, right: int, value: T
+    ) -> None:
+        """
+        Helper method to update a range of values in the segment tree.
+
+        Args:
+            node: Current node index
+            start: Start index of the current segment
+            end: End index of the current segment
+            left: Start index of the range to update
+            right: End index of the range to update
+            value: Value to apply to the range
+        """
+        # Push lazy updates before processing this node
+        self._push_lazy(node, start, end)
+
+        # No overlap
+        if start > right or end < left:
+            return
+
+        # Complete overlap
+        if start >= left and end <= right:
+            # Apply update to current node
+            self.tree[node] = self.update_op(self.tree[node], value)
+
+            # If not leaf, mark children for lazy update
+            if start != end:
+                left_child = 2 * node
+                right_child = 2 * node + 1
+
+                lazy_left_child = self.lazy[left_child]
+                if lazy_left_child is None:
+                    self.lazy[left_child] = value
+                else:
+                    self.lazy[left_child] = self.update_op(lazy_left_child, value)
+
+                lazy_right_child = self.lazy[right_child]
+                if lazy_right_child is None:
+                    self.lazy[right_child] = value
+                else:
+                    self.lazy[right_child] = self.update_op(lazy_right_child, value)
```

Comment:
nit: for loop ? also, this seems the same as lines 87-104 above. refactor ?

Reply:
done